Fixes conditional jump in handleClientEvent and cleans up sockets properly when in bad state.

Haliris opened the pull request:
Adds: infoStatus int storing return values of getpeername() and getsockname() in handleClientEvent(). This int is used to check for sockets in bad states, if the check fails, then we clean up the sockets as if the client had disconnected cleanly.

Fixes: valgrind errors when pressing down f5 in browser, but more generally, allows the  server to clean up bad sockets independently of what created them.

Commit:
Added check for return value of getpeer/sockname()

Tweak: Added branch for early return if getpeer() or getsockname() fail.

Fix: Conditional jump due to unpopulated arrays in case of getpeer() or
getsockname() failure

Commit:
Merge branch 'main' into conditional_fix

Commit:
Adds socket cleanup to getsock/peername failure branch

Tweak: Added int infoStatus to keep track of the two functions return
values and cleans up socket and partial request buffer if needed before
returning early from handleClientEvent

Fix: Server leaking like 8 megabytes and reaching its fd system limit
whenever new client connections would be opened with the previous ones
not being closed properly, like in the situation of holding down F5 in a
browser

dpblue commented:
Ok parfait pour l'ajout de SO_KEEPALIVE et pour le test des valeurs de retour de getpeername() and getsockname().
Par contre, la partie cleanup n'est-elle  pas rendondante avec le cleanup qui se fait au niveau du traitement des events ?
En effet, si on laisse le code se poursuivre, l'event va "correctement fail" avec des throw catchés et le cleanup qui va bien au niveau de safeRecv() pour l'event EPOLLIN, de handleOutgoingData() pour l'event EPOLLOUT et naturellement dans EPOLLRDUP | EPOLLUP.

dpblue commented:
Je suggère de valider en mettant en commentaire la partie cleanup et de la supprimer plus tard si l'on constate que tout se passe bien au niveau des events (ce qui me semble etre le cas actuellement). Ca fait moins de code à maintenir et less is more !

Haliris commented:
> Ok parfait pour l'ajout de SO_KEEPALIVE et pour le test des valeurs de retour de getpeername() and getsockname(). Par contre, la partie cleanup n'est-elle pas rendondante avec le cleanup qui se fait au niveau du traitement des events ? En effet, si on laisse le code se poursuivre, l'event va "correctement fail" avec des throw catchés et le cleanup qui va bien au niveau de safeRecv() pour l'event EPOLLIN, de handleOutgoingData() pour l'event EPOLLOUT et naturellement dans EPOLLRDUP | EPOLLUP.

On return tot car sinon valgrind lance une erreur de conditional jump etant donne que les tableaux ne sont pas correctement initialises.
Sinon oui c'est totalement redondant.

dpblue commented:
Il faudrait changer le ligne 101 de Server.cpp pour :
ssize_t bytes_sent = send(clientSocket, toSend->getbufptr(), toSend->getbuflen(), MSG_NOSIGNAL);

Commit:
Removes redundant cleanup logic, now protects problem functions

Tweak: Swapped send() call with ssize_t bytes_sent = send(clientSocket, toSend->getbufptr(), toSend->getbuflen(), MSG_NOSIGNAL);

Tweak: Removed redundant cleanup logic, now the functions as well as the debug prints are protected behind the
infoStatus check

dpblue commented:
LGTM !

dpblue reviewed (approved):
Good !
